src/main/java/org/spongepowered/api/tag/Tag.java

```diff
@@ -0,0 +1,117 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.tag;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.ResourceKeyed;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.Key;
+import org.spongepowered.api.registry.DefaultedRegistryValue;
+
+import java.util.Collection;
+
+/**
+ * Represents a Tag for a given type.
+ * Tags group {@link TagTypes}.
+ *
+ * @param <T> Type that this tag groups
+ */
+public interface Tag<T> extends DefaultedRegistryValue, ResourceKeyed, TagRegistration {
+
+    /**
+     * Creates a builder to create a new {@link Tag}
+     * or modify an existing one.
+     * @return Builder to build a new Tag.
+     */
+    @SuppressWarnings("unchecked")
+    static Builder<?> builder() {
+        return Sponge.game().builderProvider().provide(Builder.class);
+    }
+
+    /**
+     * Get every value in this Tag.
+     *
+     * @return All tag values
+     */
+    Collection<T> all();
+
+    interface Builder<T extends Taggable> extends org.spongepowered.api.util.ResourceKeyedBuilder<TagRegistration, Builder<T>> {
+
+        /**
+         * Sets the tag type and generifies the builder.
+         *
+         * @param tagType Type to build a tag for, e.g {@link TagTypes#BLOCK_TYPE}
+         * @return This builder, for chaining
+         */
+        <NT extends Taggable> Builder<NT> type(TagType<NT> tagType);
+
+        /**
+         * <p>Whether to replace instead of append if the tag already
+         * exists.</p>
+         *
+         * <p>By default, if the {@link ResourceKey#value()} is the same as
+         * another data pack / plugin, the two will be combined, and their
+         * tags appended to each other.</p>
+         *
+         * @param replace Whether to replace instead of merge.
+         * @return This builder, for chaining.
+         */
+        Builder<T> replace(boolean replace);
+
+        default Builder<T> replace() {
+            return replace(true);
+        }
+
+        /**
+         * Adds a child tag to this tag.
+         * This adds all of its values to this tag.
+         *
+         * @param childTag Child tag to add.
+         * @return This builder, for chaining
+         */
+        Builder<T> addChild(Tag<T> childTag);
+
+        Builder<T> addChildren(Collection<Tag<T>> children);
+
+        /**
+         * Adds the value to the tag.
+         *
+         * @param value Value to add
+         * @return This builder, for chaining
+         */
+        Builder<T> addValue(T value);
+
+        Builder<T> addValues(Collection<T> values);
+
+        /**
+         * A nicely generified tag, <b>That must be registered</b> in {@link org.spongepowered.api.event.lifecycle.RegisterDataPackValueEvent}.
+         *
+         * @return The built tag.
+         */
+        @Override
+        @NonNull TagRegistration build();
```

Comment:
...so what's the point of the separation between a `TagRegistration` and a `Tag`?

Also, we're not registering the tag at this point, right? So this isn't a registration...

Reply:
A TagRegistration is representative of the actual json file. TagRegistration is a addition or replacement of tags by a particular plugin. A tag has all of these combined, into a useable object.
I called it TagRegistration to make it like RecipeRegistration - which also hasn't been registered yet. Maybe a change of that too is in order?

Reply:
Registration because we `RegisterDataPackValueEvent#register` those `DataPackSerializable`.

In case of recipes I mirrored how vanilla creates its recipes. 
`net.minecraft.data.recipes.FinishedRecipe` = `RecipeRegistration` which knows how to serialize itself
`net.minecraft.world.item.crafting.Recipe` = `Recipe` which knows how to craft/cook etc.

For Tags we also need this distinction because datapacks can append to or replace existing tags.
What a plugin registers in its datapack is not necessarily the final state when all datapacks are loaded.

Reply:
Well, that's _not a registration_. It's not _registered_ at that point which is what my point was there.

However, my bigger point is that these things are not clear. I have delved into the source code to understand what is going on - this is how I understand it:

* A `Tag` is the basic "here is a category and you can get everything that fits in this category from this". My guess is that it is the immutable in-game representation. In general, you won't create this.
* A `TagRegistration` is "here is how and what you register". It just so happens to be a superset of what `Tag` is.

Have I got this right here? If so, then this builder needs to be an inner interface to the `TagRegistration`, and the `register` method should be exposed on said registration because the behaviour should be knowable. (On `Tag`, that wouldn't matter because it has no futher role to play in registration so the `register` method is useless).

The docs need to be clearer as to what they do too.

Reply:
@dualspiral I am likely going to institute that all the `data pack` like classes become `XXXXTemplate` as that is what they are and likely this will jive a lot better in your head then.

Observe how `WorldTypeTemplate` and `WorldType` exist and a `WorldType` may make a `WorldTypeTemplate` but a template only, eventually, becomes a type after the `RegisterDataPackValueEvent`

Reply:
also keep in mind that Datapacks can be reloaded and so the plugin must register it every time. `RegisterDataPackValueEvent` is called

Reply:
> @dualspiral I am likely going to institute that all the `data pack` like classes become `XXXXTemplate` as that is what they are and likely this will jive a lot better in your head then.
> 
> Observe how `WorldTypeTemplate` and `WorldType` exist and a `WorldType` may make a `WorldTypeTemplate` but a template only, eventually, becomes a type after the `RegisterDataPackValueEvent`

Sensible enough to me. Consistency is important. In that regard I'm not sure if `TagTemplate` or whatever should implement `Tag` altogether, if we're going to make that distinction.

> also keep in mind that Datapacks can be reloaded and so the plugin must register it every time. `RegisterDataPackValueEvent` is called

Right - but that doesn't change anything here.

Reply:
@dualspiral You are correct. a `TagTemplate` eventually transforms into a `Tag` but they are definitely not the same thing.

@Faithcaio Can this happen the same for recipes? Consistency as dual put it is very important. For plugin devs, they should know that they build `RecipeTemplates` and submit it to our event. Getting the actual `Recipe` later is looking into the registry system.

Basically the `data pack template` classes are immutable things you build once, discard, Vanilla/Sponge ops on it to write to filesystem then discarded.

Reply:
Not exactly the same but yes thats how they work.

It is not consistent with Advancements.

Reply:
In that vein then, `AdvancementTemplates` should be a thing, separated from `Advancement`.